cephfs-top: addition of sort feature and limit option

neesingh-rh opened the pull request:
This PR intends to add two new features to cephfs-top:
* sort-by field value
* limit the number of clients to be displayed

Fixes: https://tracker.ceph.com/issues/55121
Signed-off-by: Neeraj Pratap Singh <neesingh@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

anthonyeleven reviewed (approved):
Docs lgtm

neesingh-rh commented on src/tools/cephfs/top/cephfs-top:
```diff
@@ -102,6 +102,8 @@ last_read_size = {}
 last_write_size = {}
 
 fs_list = []
+sort_dict = {"last_fs": "", "last_field": 'chit', "limit": None}
```
No, it is juat a data structure to store some states.

vshankar reviewed (commented):
@neesingh-rh I'm playing with this from the last couple of days. Works really well :thumbsup:

Really nice work!

vshankar commented:
@joscollin please review this change.

joscollin commented:
@vshankar Looks good at the first look. Will review more tomorrow. 
Just FYI, rbd-top has a smart way of sort-by.

vshankar commented:
> @vshankar Looks good at the first look. Will review more tomorrow. Just FYI, rbd-top has a smart way of sort-by.

Yeh, I know that - the sort limit field can be specified when registering the query with OSD. That's supported by the MDS too :)

vshankar commented:
> @vshankar Looks good at the first look. Will review more tomorrow. Just FYI, rbd-top has a smart way of sort-by.

@joscollin Is this good to go?

vshankar commented:
jenkins test make check

joscollin commented:
> > @vshankar Looks good at the first look. Will review more tomorrow. Just FYI, rbd-top has a smart way of sort-by.
> 
> @joscollin Is this good to go?

Will review it again the next working day.

anthonyeleven commented on doc/cephfs/cephfs-top.rst:
```diff
@@ -86,7 +86,12 @@ Interactive Commands
+      Allows to choose/change the below options:
+      * sort:  The 's' key designates a field as the sort field, 'cap_hit' is the default sort order.
+      * limit: The 'l' key lets you set the limit on the number of clients to be displayed.
+
```
Suggest:

```
. f : Field management
      * sorting:  The 's' key designates the sort field.  'cap_hit' is the default.
      * client limit: The 'l' key sets the limit on the number of clients to be displayed.
```

neesingh-rh replied:
Done.

joscollin reviewed (changes requested):
Some limit menu issues:
1. Launch the limit menu, resize the window multiple times, the menu is gone.
2. Launch the limit menu, input a number and then resize the window, the menu is gone.
3. Launch the limit menu, press any random keys, the menu is gone.
4. Why the inputed number is not displayed on the screen?
5. Launch the limit menu, pressing 'q' doesn't immediately go back to the previous screen. It needs multiple q presses it seems.

Is there a need for 'Fields Management' menu screen, as it just a redirect to another selection screen? We could do the same from the home page?
Like:
Press 'm' to select a filesystem, 's' for sort, 'l' to limit the clients and  and 'q' to quit

@vshankar @gregsfortytwo 
Sort is fine, but do we really need a limit as it could be scrolled when #48090 is merged? Looks like once the limit is set, it cannot be reset. We need to exit cephfs-top and launch again.

joscollin commented:
I think it would be better have something like a 'Reset to default' key at the home screen itself. So that the user can reset to the default home screen view anytime, without going inside the Sort screen and select the 'default'. This new key would reset the current sort & limit setting.

vshankar commented:
> Some limit menu issues:
> 
> 1. Launch the limit menu, resize the window multiple times, the menu is gone.
> 2. Launch the limit menu, input a number and then resize the window, the menu is gone.
> 3. Launch the limit menu, press any random keys, the menu is gone.
> 4. Why the inputed number is not displayed on the screen?
> 5. Launch the limit menu, pressing 'q' doesn't immediately go back to the previous screen. It needs multiple q presses it seems.
> 
> Is there a need for 'Fields Management' menu screen, as it just a redirect to another selection screen? We could do the same from the home page? Like: Press 'm' to select a filesystem, 's' for sort, 'l' to limit the clients and and 'q' to quit
> 
> @vshankar @gregsfortytwo Sort is fine, but do we really need a limit as it could be scrolled when #48090 is merged? Looks like once the limit is set, it cannot be reset. We need to exit cephfs-top and launch again.

I still find the limit functionality useful - think of using it with sort and limiting the display to top two clients that are consuming bandwidth. As far as resetting the limit, I think that a bug and should be fixed.

vshankar commented:
@neesingh-rh 

```
cephfs-top:318:101: E501 line too long (110 > 100 characters)
cephfs-top:321:101: E501 line too long (103 > 100 characters)
cephfs-top:322:101: E501 line too long (115 > 100 characters)
cephfs-top:401:101: E501 line too long (117 > 100 characters)
cephfs-top:402:21: E128 continuation line under-indented for visual indent
cephfs-top:772:101: E501 line too long (160 > 100 characters)
cephfs-top:776:21: E117 over-indented
cephfs-top:777:101: E501 line too long (145 > 100 characters)
cephfs-top:904:101: E501 line too long (109 > 100 characters)
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/cephfs/top/.tox/py3/bin/flake8 --ignore=W503 --max-line-length=100 cephfs-top (exited with code 1)
```

neesingh-rh commented:
> Some limit menu issues:
> 
>     1. Launch the limit menu, resize the window multiple times, the menu is gone.
> 
>     2. Launch the limit menu, input a number and then resize the window, the menu is gone.
> 
>     3. Launch the limit menu, press any random keys, the menu is gone.
> 
Above issues are resolved, RESIZE was also accepted as input earlier.
>     4. Why the inputed number is not displayed on the screen?
> 
Added this functionality.
>     5. Launch the limit menu, pressing 'q' doesn't immediately go back to the previous screen. It needs multiple q presses it seems.
> 
Fixed, Earlier it was working by pressing Enter immediately after pressing 'q'.
> 
> Is there a need for 'Fields Management' menu screen, as it just a redirect to another selection screen? We could do the same from the home page? Like: Press 'm' to select a filesystem, 's' for sort, 'l' to limit the clients and and 'q' to quit
> 
I don't want to remove the Fields Management. It is keeping the things separated and easy to handle, even when some new functionalities being added later.And follows top(1) structure too.
> @vshankar @gregsfortytwo Sort is fine, but do we really need a limit as it could be scrolled when #48090 is merged? Looks like once the limit is set, it cannot be reset. We need to exit cephfs-top and launch again.
It wasn't the bug earlier, we could set the limit value as many times we want, "d" key was there to reset. Now, I have added BACKSPACE too, to change the limit value during input time and currently I have set the length to digits allowed is FOUR.

neesingh-rh commented:
> I think it would be better have something like a 'Reset to default' key at the home screen itself. So that the user can reset to the default home screen view anytime, without going inside the Sort screen and select the 'default'. This new key would reset the current sort & limit setting.

I think its better to have the default "d" key added to FIELDS MANAGEMENT Screen, which will makes the resetting the FIELD MANAGEMENT values only. Its good to have things modularized. I have added "d" key on the FIELDS MANAGEMNT Screen, it will reset the sort as well as limit value to the default.

neesingh-rh commented on src/tools/cephfs/top/cephfs-top:
```diff
@@ -270,22 +292,141 @@ class FSTop(object):
+                      "rsp= READ_IO_SPEED", "wtio= WRITE_IO_SIZES", "waio= WRITE_AVG_IO_SIZES",
+                      "wsp= WRITE_IO_SPEED", "rlatavg= AVG_READ_LATENCY", "rlatsd= STDEV_READ_LATENCY",
+                      "wlatavg= AVG_WRITE_LATENCY", "wlatsd= STDEV_WRITE_LATENCY", "mlatavg= AVG_METADATA_LATENCY",
+                      "mlatsd= STDEV_METADATA_LATENCY", "Default"]
```
Firstly, I tried using that only for easy maintenance but it wasn't successful. Because currently we are using the field names stored in the `field_menu` to match for the field chosen and the display structure of field menu is followed the way it is done in top(1). If I use the `MAIN_WINDOW_TOP_LINE_METRICS` , firstly I will have to call its respective function for eg `self.items()` and self.avg_items()` again and again.Secondly, `MAIN_TOP_LINE_METRICS` have some fields which are not being displayed directly in cephfs-top display as fields, their avg, std is being used but for us, order is very important.That's why, it looked better and easy to do this way.

neesingh-rh commented:
> > Some limit menu issues:
> > 
> > 1. Launch the limit menu, resize the window multiple times, the menu is gone.
> > 2. Launch the limit menu, input a number and then resize the window, the menu is gone.
> > 3. Launch the limit menu, press any random keys, the menu is gone.
> > 4. Why the inputed number is not displayed on the screen?
> > 5. Launch the limit menu, pressing 'q' doesn't immediately go back to the previous screen. It needs multiple q presses it seems.
> > 
> > Is there a need for 'Fields Management' menu screen, as it just a redirect to another selection screen? We could do the same from the home page? Like: Press 'm' to select a filesystem, 's' for sort, 'l' to limit the clients and and 'q' to quit
> > @vshankar @gregsfortytwo Sort is fine, but do we really need a limit as it could be scrolled when #48090 is merged? Looks like once the limit is set, it cannot be reset. We need to exit cephfs-top and launch again.
> 
> I still find the limit functionality useful - think of using it with sort and limiting the display to top two clients that are consuming bandwidth. As far as resetting the limit, I think that a bug and should be fixed.

+1 , Scroll makes the user to have a view of all the clients easily but if user wants to analyze metrics for some top 50 or 60 clients among thousands of clients, he or she will have an option.

neesingh-rh commented:
> @neesingh-rh
> 
> ```
> cephfs-top:318:101: E501 line too long (110 > 100 characters)
> cephfs-top:321:101: E501 line too long (103 > 100 characters)
> cephfs-top:322:101: E501 line too long (115 > 100 characters)
> cephfs-top:401:101: E501 line too long (117 > 100 characters)
> cephfs-top:402:21: E128 continuation line under-indented for visual indent
> cephfs-top:772:101: E501 line too long (160 > 100 characters)
> cephfs-top:776:21: E117 over-indented
> cephfs-top:777:101: E501 line too long (145 > 100 characters)
> cephfs-top:904:101: E501 line too long (109 > 100 characters)
> ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/cephfs/top/.tox/py3/bin/flake8 --ignore=W503 --max-line-length=100 cephfs-top (exited with code 1)
> ```

Removed flake8 errors. Thanks

joscollin commented:
> > > Some limit menu issues:
> > > 
> > > 1. Launch the limit menu, resize the window multiple times, the menu is gone.
> > > 2. Launch the limit menu, input a number and then resize the window, the menu is gone.
> > > 3. Launch the limit menu, press any random keys, the menu is gone.
> > > 4. Why the inputed number is not displayed on the screen?
> > > 5. Launch the limit menu, pressing 'q' doesn't immediately go back to the previous screen. It needs multiple q presses it seems.
> > > 
> > > Is there a need for 'Fields Management' menu screen, as it just a redirect to another selection screen? We could do the same from the home page? Like: Press 'm' to select a filesystem, 's' for sort, 'l' to limit the clients and and 'q' to quit
> > > @vshankar @gregsfortytwo Sort is fine, but do we really need a limit as it could be scrolled when #48090 is merged? Looks like once the limit is set, it cannot be reset. We need to exit cephfs-top and launch again.
> > 
> > 
> > I still find the limit functionality useful - think of using it with sort and limiting the display to top two clients that are consuming bandwidth. As far as resetting the limit, I think that a bug and should be fixed.
> 
> +1 , Scroll makes the user to have a view of all the clients easily but if user wants to analyze metrics for some top 50 or 60 clients among thousands of clients, he or she will have an option.

I don't see an adequate reason why a user does so. It still doesn't make any betterment than just top. If `limit` is an option to export-data/interface to other tools, then yes there is a need. Otherwise, it looks like a confusing user interface to me. Anyway, I'd say let's keep some confusion for the user to play with :)

joscollin commented:
> > Some limit menu issues:
> > ```
> > 1. Launch the limit menu, resize the window multiple times, the menu is gone.
> > 
> > 2. Launch the limit menu, input a number and then resize the window, the menu is gone.
> > 
> > 3. Launch the limit menu, press any random keys, the menu is gone.
> > ```
> 
> Above issues are resolved, RESIZE was also accepted as input earlier.
> 
> > ```
> > 4. Why the inputed number is not displayed on the screen?
> > ```
> 
> Added this functionality.
> 
> > ```
> > 5. Launch the limit menu, pressing 'q' doesn't immediately go back to the previous screen. It needs multiple q presses it seems.
> > ```
> 
> Fixed, Earlier it was working by pressing Enter immediately after pressing 'q'.
> 
> > Is there a need for 'Fields Management' menu screen, as it just a redirect to another selection screen? We could do the same from the home page? Like: Press 'm' to select a filesystem, 's' for sort, 'l' to limit the clients and and 'q' to quit
> 
> I don't want to remove the Fields Management. It is keeping the things separated and easy to handle, even when some new functionalities being added later.And follows top(1) structure too.

I've never been impressed with top(1)'s user interface honestly. The scroll feature is unusable in top(1) sometimes (just try that) and the other selection screens are confusing to the user too. I think the latest cephfs-top user interface is way better than top(1). So I don't look at top(1) anymore.

But after using the Fields Management screen in cephfs-top, I see that it's annoying the user by prompting the option repeatedly. So it would be easily accessible from the home screen. When there are several new options (in the future), I'd prefer some kind of drop down menu (or a similar thing) from the home screen for a better arrangement, which I've been looking at a few days ago.

vshankar commented:
LGTM. @joscollin Need your approval for merge.

joscollin reviewed (changes requested):
Looks really nice now. I have a few comments:

1. Do we really need a `press "d" for default` option inside the `l` menu? As default d is already there in the home screen, the user comes inside only for a custom value.
2. I think we certainly don't need filter to `0` clients. Please avoid that.
3. Doc screenshot update?
4. In the `l` menu, `q` quits only when the field is blank. I think we should enable `q` always in-case the user changes his mind.
5. In the home screen, change to `Press 'd' for default setting`. 
6. The `HELP` is getting bigger every-time. How about shortening it a little bit like: `COMMANDS: m - select a filesystem | s - sort menu | l - limit number of clients | d - default setting | q - quit` ?

Will check the code tomorrow.

neesingh-rh commented:
> Looks really nice now. I have a few comments:
> 
>     1. Do we really need a `press "d" for default` option inside the `l` menu? As default d is already there in the home screen, the user comes inside only for a custom value.
>
Let's keep it this way. If user wants to default value only for limit not for the sort value, can explicitly do so. 
>     2. I think we certainly don't need filter to `0` clients. Please avoid that.
> 
Done.
>     3. Doc screenshot update?
> 
Added
>     4. In the `l` menu, `q` quits only when the field is blank. I think we should enable `q` always in-case the user changes his mind.
> 
Enabled 'q' while giving input too.
>     5. In the home screen, change to `Press 'd' for default setting`.
> 
updated
>     6. The `HELP` is getting bigger every-time. How about shortening it a little bit like: `COMMANDS: m - select a filesystem | s - sort menu | l - limit number of clients | d - default setting | q - quit` ?
> 
Done
> 
> Will check the code tomorrow.

joscollin commented:
> > Looks really nice now. I have a few comments:
> > ```
> > 1. Do we really need a `press "d" for default` option inside the `l` menu? As default d is already there in the home screen, the user comes inside only for a custom value.
> > ```
> 
> Let's keep it as the existing way only. If user wants default value only for limit not for the sort value, can explicitly do so.
> 

Then mention that feature in the doc under ` d : Default`.

neesingh-rh commented on src/tools/cephfs/top/cephfs-top:
```diff
@@ -764,11 +965,19 @@ class FSTop(object):
+                        num_client = int(current_states['limit'])
+                    else:
+                        num_client = len(client_metadata)
+                    vscrollEnd += num_client
```
Fixed.

joscollin commented on src/tools/cephfs/top/cephfs-top:
```diff
@@ -270,26 +301,171 @@ class FSTop(object):
+                    if isinstance(last_fs, list):
+                        self.run_all_display()
+                    else:
+                        self.run_display()
```
I think this could be simplified by making the the below code common for both cases.

```
                    if isinstance(last_fs, list):
                        self.run_all_display()
                    else:
                        self.run_display()
```

neesingh-rh replied:
Yup, Thanks

joscollin commented:
@neesingh-rh Fix the tox failures too.

joscollin commented on src/tools/cephfs/top/cephfs-top:
```diff
@@ -270,26 +301,171 @@ class FSTop(object):
+                    self.run_all_display()
+                else:
+                    self.run_display()
+
```
This also crashes when the window is resized after the limit number input and before the enter key is pressed.

neesingh-rh replied:
Fixed.

joscollin commented on src/tools/cephfs/top/cephfs-top:
```diff
@@ -270,26 +301,171 @@ class FSTop(object):
             curses.halfdelay(self.refresh_interval_secs)
             key = stdscr.getch()
 
+    def set_limit(self, stdscr):
```
The `set_limit` code could be simplified by using a `Textbox` inside a `rectangle` for input. But let's do that later, not in this PR.

joscollin reviewed (approved):
qa doesn't seems necessary (it's just a selftest). Let's go for merge.

joscollin commented:
@vshankar 

vstart_runner test succeed:

```
2022-11-04 07:04:17,323.323 INFO:tasks.cephfs.fuse_mount:Found client admin socket at /home/jcollin/workspace/ceph/build/asok/client.0.1664051.asok
2022-11-04 07:04:17,323.323 DEBUG:__main__:"sudo" was omitted from the following cmd args before execution and logging using function overriding; check vstart_runner.py for more details.
2022-11-04 07:04:17,323.323 DEBUG:__main__:> sudo ./bin/ceph --admin-daemon /home/jcollin/workspace/ceph/build/asok/client.0.1664051.asok status
2022-11-04 07:04:17,425.425 DEBUG:__main__:> stat --file-system '--printf=%T
' -- /tmp/tmpj430z5eb/mnt.0
2022-11-04 07:04:17,427.427 INFO:tasks.cephfs.fuse_mount:ceph-fuse is mounted on /tmp/tmpj430z5eb/mnt.0
2022-11-04 07:04:17,427.427 DEBUG:__main__:> sudo chmod 1777 /tmp/tmpj430z5eb/mnt.0
2022-11-04 07:04:17,436.436 DEBUG:__main__:> ./bin/ceph mgr module enable stats
2022-11-04 07:04:18,679.679 DEBUG:__main__:> /home/jcollin/workspace/ceph/src/tools/cephfs/top/cephfs-top --cluster=hpec --id=admin --selftest
cluster hpec does not exist
2022-11-04 07:04:18,755.755 DEBUG:__main__:> ./bin/ceph mgr module disable stats
2022-11-04 07:04:19,798.798 DEBUG:__main__:> set -ex
dd if=/proc/self/mounts of=/dev/stdout
2022-11-04 07:04:19,799.799 DEBUG:__main__:Discovered MDS IDs: dict_keys(['a', 'b', 'c'])
2022-11-04 07:04:19,799.799 DEBUG:__main__:> ./bin/ceph osd blocklist ls
2022-11-04 07:04:20,039.039 INFO:tasks.cephfs.fuse_mount:Running fusermount -u on local...
2022-11-04 07:04:20,039.039 DEBUG:__main__:> sudo fusermount -u /tmp/tmpj430z5eb/mnt.0
2022-11-04 07:04:20,048.048 INFO:tasks.cephfs.mount:Cleaning up mount local
2022-11-04 07:04:20,048.048 DEBUG:__main__:> rmdir -- /tmp/tmpj430z5eb/mnt.0
2022-11-04 07:04:20,049.049 DEBUG:__main__:> ./bin/ceph fs dump --format=json
2022-11-04 07:04:20,307.307 DEBUG:__main__:Discovered MDS IDs: dict_keys(['a', 'b', 'c'])
2022-11-04 07:04:20,307.307 DEBUG:__main__:> ./bin/ceph fs dump --format=json
2022-11-04 07:04:20,579.579 DEBUG:__main__:> ./bin/ceph osd dump --format=json
2022-11-04 07:04:20,851.851 INFO:tasks.cephfs.filesystem:Destroying file system cephfs and related pools
2022-11-04 07:04:20,851.851 DEBUG:__main__:> ./bin/ceph fs dump --format=json
2022-11-04 07:04:21,115.115 DEBUG:__main__:> ./bin/ceph fs dump --format=json
2022-11-04 07:04:21,372.372 DEBUG:__main__:> ./bin/ceph osd dump --format=json
2022-11-04 07:04:21,606.606 DEBUG:__main__:> ./bin/ceph fs fail cephfs
2022-11-04 07:04:22,984.984 DEBUG:__main__:> ./bin/ceph fs rm cephfs --yes-i-really-mean-it
2022-11-04 07:04:24,008.008 DEBUG:__main__:> ./bin/ceph osd pool rm cephfs_metadata cephfs_metadata --yes-i-really-really-mean-it
2022-11-04 07:04:25,144.144 DEBUG:__main__:> ./bin/ceph osd pool rm cephfs_data cephfs_data --yes-i-really-really-mean-it
2022-11-04 07:04:26,283.283 DEBUG:__main__:> ./bin/ceph config rm mon mon_allow_pool_delete
2022-11-04 07:04:26,552.552 DEBUG:__main__:> ./bin/ceph log 'Ended test tasks.cephfs.test_fstop.TestFSTop.test_fstop_non_existent_cluster'
2022-11-04 07:04:27,229.229 INFO:__main__:Stopped test: test_fstop_non_existent_cluster (tasks.cephfs.test_fstop.TestFSTop) in 28.025532s
2022-11-04 07:04:27,229.229 INFO:__main__:test_fstop (tasks.cephfs.test_fstop.TestFSTop) ... ok
2022-11-04 07:04:27,229.229 INFO:__main__:test_fstop_non_existent_cluster (tasks.cephfs.test_fstop.TestFSTop) ... ok
2022-11-04 07:04:27,229.229 INFO:__main__:
2022-11-04 07:04:27,229.229 INFO:__main__:----------------------------------------------------------------------
2022-11-04 07:04:27,229.229 INFO:__main__:Ran 2 tests in 71.518s
2022-11-04 07:04:27,230.230 INFO:__main__:
2022-11-04 07:04:27,230.230 INFO:__main__:
2022-11-04 07:04:27,230.230 DEBUG:__main__:> ip netns list
2022-11-04 07:04:27,231.231 DEBUG:__main__:> sudo ip link delete ceph-brx
Cannot find device "ceph-brx"
2022-11-04 07:04:27,243.243 INFO:__main__:OK
2022-11-04 07:04:27,243.243 INFO:__main__:

```
